### ISSUE TITLE
Create Standalone ZoomPanLayout (better name?)

### ISSUE THREAD

**Issue by peterLaurence** [moagrius/TileView#418]

**Comment by moagrius** [comment]
I think this should be a subclass of a 2d scroller...

**Comment by peterLaurence** [comment]
Nice you're there. I was thinking about that.
So i agree we could find a better name.
But i'm not sure i understand the purpose of of this class extending a Scroller. Could you tell me?

**Comment by peterLaurence** [comment]
The javadoc of Scroller shows the typical use : a given class bundles a Scroller, and then `mScroller.computeScrollOffset()` is used in the View's `computeScroll` to set scroll positions. This is what is done in ZoomPanLayout. 
I just need to understand the benefit of having the Layout (or ViewGroup) being a Scroller.

**Comment by peterLaurence** [comment]
About the name : i propose SurfaceLayout

**Comment by moagrius** [comment]
The scroller would be for lots of things - e.g., a big table or grid of views, or just a bunch of statically positioned views larger than the screen - there are some open source attempts out there - it's just a 2d scroller that doesn't have scaling and pinching. 

The next level down would be what we now call ZoomPanLayout - basically 2d scroll with pinch and zoom (but the implementer happens what actually happens with the scale).

SurfaceView already exists - it's a View you draw textures on (like video)

**Comment by peterLaurence** [comment]
Ok then. That 2D Scroller would be raw Scroller, or a View having a Scroller instance?

**Comment by peterLaurence** [comment]
Have you already started?
I could go ahead. There is one thing we need to think about first : the ZoomPanLayout already extends ViewGroup. It won't be able to extend another class.

**Comment by moagrius** [comment]
Dangit, I didn't get this last notice.

> Have you already started?

No I haven't started.

> There is one thing we need to think about first : the ZoomPanLayout already extends ViewGroup. It won't be able to extend another class.

Right, there's going to be a lot changes.  I'd first do a 2D scroller, with an API similar to `ScrollView`, that takes one child, and scrolls that child (just like a ScrollView).

Once that's done, then ZoomPanLayout (or whatever it ends up being called) can inherit from that 2d ScrollView...

Does that make sense?

**Comment by peterLaurence** [comment]
That makes sense. Too bad its now late for me, but i will be able start something tomorrow.

**Comment by moagrius** [comment]
Gotcha.  No rush!

I'll try to keep an eye out this weekend for anything you post.

Thanks, Peter

**Comment by peterLaurence** [comment]
I pushed on my fork, v3 branch : https://github.com/peterLaurence/TileView/commit/58f2605a077c47a1cfe51ead353fffa9c51c725e

It's a refactor, with the following changes :
1. The 2D Scroller is named View2DScroller (name inspired by [ViewPager](https://android.googlesource.com/platform/frameworks/support/+/jb-dev/v4/java/android/support/v4/view/ViewPager.java), code advertized as example of custom scrolling implementation in Android doc). It also extends ViewGroup btw.

2. In View2DScroller, calls for the scroller no longer uses a getter, to reduce overhead. As consequence, a new api `setScroller` is added to keep the ability to change of scroller.

3. Same for ZoomPanAnimator, the name is kept as is for instance.

4. Revisited javadoc, to use `{@link ClassName}`, to ease future refactor-rename.

Ofc, tested and its working with the demo.

**Comment by moagrius** [comment]
Hey Peter,

Thanks for doing that!  So I'm sure it's great, but let's step back a bit.

First, I'd like to start from scratch.  We start by thinking about the various components and APIs, and what we want to offer, and how we get there.

Without even thinking about TileView (but obviously leveraging anything we have that's successful), we come up with that first widget, in a totally new, empty project.

For example, the ScrollView should be a standalone widget.  It'll probably get it's own repo, it's own package, and won't be attached to TileView necessarily (at the start, at least, but of course TileView will probably be a subclass of this ScrollView, or a subclass of a subclass).

For the initial version of ScrollView, I think we should look at the existing (single-axis) android ScrollView and try to copy that API as much as possible, but of course leverage the code we have in ZoomPanLayout to get there.  I think listeners, methods, etc should be as familiar as possible to users of a normal Android ScrollView.

Once that's done and tested and has a working demo, we move on to "ZoomView" (terrible name I know, just for now).  ZoomView will inherit from ScrollView (or whatever name), but will also offer pinch, double-tap and programmatic zoom/scale.

Once ZoomView is done and tested and has a working demo, we move on to a basic TileView, which just extends ZoomView and takes plugins.  One plugin, for example (maybe default) will support image Tilings.  Once that's done, tested, etc - we move on to the next plugin, maybe Markers, then maybe Paths, etc.

Someone should be able to use the ScrollView without having to use TileView someone should be able to use ZoomView without TileView.  Etc.

I'm not insisting that we do anything in any particular way, but one thing I've found very helpful in recent years, is to write out a class's entire API in "english" before coding.

Also, that style was one I was using a long time ago - I'm very happy to use normal Android conventions now - I think the style we used in TileView is mostly OK, minus the extra whitespace.  We use a very standard one at work, it be very handy to use the same for this project - I can upload the XML later.

Does that all make sense?  What are you thoughts at this point?

**Comment by peterLaurence** [comment]
Oh i didn't realize you wanted to create sister projects.
So why not. I like your idea. I just hope that this won't be to the detriment of the efficiency of the library (e.g avoid nested layouts). But i'm not too worried because as you said TileView extends ZoomView which extends ScrollView (so no nested layouts).
One other thing bugs me. Why the hell there is no official 2D ScrollView or at least a very well know library? I find this very suspicious.
But i don't see why this wouldn't possible, you've already done that (it was just mixed with the zoom feature).

So ok for writing down the ScrollView api (we should really find a definitive name). I see it as a lightweight version of the official ScrollView (with a Scroller instead of OverScroller for example), and with the horizontal scroll capability.

**Comment by moagrius** [comment]
>  I just hope that this won't be to the detriment of the efficiency of the library (e.g avoid nested layouts)

Yeah there won't be any issue - honestly nested layouts isn't where most of our danger is, but you're right it should be fine - I don't imagine there will be any more nesting than we have now.

> One other thing bugs me. Why the hell there is no official 2D ScrollView or at least a very well know library? I find this very suspicious.

I agree, and that's one reason why I think this is a good opportunity to provide one.  At work the other day we wanted to just blow up an image when it was tapped - I found a few that did this, but most had either really unfriendly APIs, or the one I ended up using (PhotoView) has 10,000 methods: http://www.methodscount.com/?lib=com.github.chrisbanes%3APhotoView%3A1.3.0.  When I first was trying to do TileView, I went looking for a 2D scrollview and never found one - in fact the only one that was close wasn't quite complete.

> So ok for writing down the ScrollView api (we should really find a definitive name). I see it as a lightweight version of the official ScrollView (with a Scroller instead of OverScroller for example), and with the horizontal scroll capability.

Yeah, I think we can really just take most of it from ZoomPanLayout, and move the methods around to more closely match the existing ScrollView API.  One thing, for example, that the "real" ScrollView does is enforce a 1 child limit - which makes onMeasure and onLayout much easier.  I can fill those in if you're not comfortable doing it.  I don't mind using a Scroller instead of an OverScroller, but I think we should give the option for an OverScroller.  Honestly, whatever the native ScrollViews does is what I'd think is appropriate - we'd want people unfamiliar with TileView at all to be able to pick up the ScrollView widget and use it just like they'd use a native ScrollView (as much as possible) - I think most of the hard work is done, just need to make it "feel" familiar.

That leads me to one other thing - the package name.  Since we want people to be able to use ScrollView or ZoomView without TileView, we should probably rename the package (and I might end up putting each of them in their own repo, but all the code would be shared - it'd repeat in TileView, for example).  Just to be simple and obvious, I see a lot of people just use their github as the package - so `com.github.moagrius` as the parent, so maybe  `com.github.moagrius.ScrollView` or `com.github.moagrius.widgets.ScrollView`...

For the name, I'm hesitant to use `ScrollView` since there's a native widget, but still the canonical name is unique...  What do you think?

**Comment by peterLaurence** [comment]
> One thing, for example, that the "real" ScrollView does is enforce a 1 child limit - which makes onMeasure and onLayout much easier. I can fill those in if you're not comfortable doing it.

For instance, i can't tell. But sure we will share the work :) 

> For the name, I'm hesitant to use ScrollView since there's a native widget, but still the canonical name is unique... What do you think?

Yes the canonical name is unique, so we can keep the ScrollView name. I'm ok with that.

**Comment by peterLaurence** [comment]
My plan for the moment is to read the official ScrollView source code, and HorizontalScrollView. Seeing the difference, we can think as a way to provide a unified api.

**Comment by moagrius** [comment]
:thumbsup:

**Comment by peterLaurence** [comment]
Both views have basically the same api, with subtle differences.
From there we could take the ScrollView api. But merging will be a massive undertaking, unless we want to simplify (there is ~1500 lines of code). 
What do you think?

**Comment by moagrius** [comment]
Yeah, I think there's only a few methods we'd have to fill in (that aren't already done in ZoomPanLayout), mostly the key event stuff.

Maybe I could try to create an initial draft sometime this week, then push it up to a new v3 branch, just as a starting point, for review, and to complete by whichever one of us has time (and probably both - we can just work in small, atomic sprints).  Sound reasonable?

**Comment by peterLaurence** [comment]
Ok :+1:

**Comment by moagrius** [comment]
Will do

**Comment by moagrius** [comment]
Hey Peter, just a note that I am working on this - I have 2 versions half-baked, one basically modifying the framework's version, and one stripping down ZoomPanLayout.  Neither are functional yet, and I'm having trouble even compiling after refactoring package names, moving files, etc.  I'm about to run out of time but will try to get something up ASAP.

**Comment by moagrius** [comment]
Hey Peter,

OK I got a first draft up.  I went with a hybrid of both the actual ScrolLView widget, but most of the actual logic is from ZoomPanLayout (although they're actually really similar).

The branch is `v3`, and since we're changing package name, I just put everything in the demo module for now: https://github.com/moagrius/TileView/tree/v3/demo/src/main/java/tileview/demo

The only "real" class is `ScrollView` https://github.com/moagrius/TileView/blob/v3/demo/src/main/java/tileview/demo/ScrollView.java (which I'd love if you could review).  I put some demos up as well if you run it (I removed the old demos).  You'll see 3 different versions - vertical, horizontal, and universal - the cool thing is you'll see there's no actual configuration changes made for the `ScrollView` - no orientation or size set - it just reacts to it's child.  I copied the public API from `android.widget.ScrollView` as much as I could, but still have a few things to do (e.g., overscroll mode).  I used the public API first:

https://developer.android.com/reference/android/widget/ScrollView.html

Then checked the source on both the oldest version (2.0) and the newest one I could find (5.1):
http://grepcode.com/file/repository.grepcode.com/java/ext/com.google.android/android/5.1.1_r1/android/widget/ScrollView.java
http://grepcode.com/file/repository.grepcode.com/java/ext/com.google.android/android/2.0_r1/android/widget/ScrollView.java/

I also compared HorizontalScrollView, and it's virtually the same code, just using width/left/right instead of height/top/bottom.

I'm thinking we move this whole thing into a new repo for now...  I created one but haven't taken the time to set it up yet.

Let me know what you think.  No rush at all.

Thanks Peter.

**Comment by peterLaurence** [comment]
Hey Mike, thanks for your feedback!

I'm glad that you got a first draft, and i'm not surprised that you kept most of the ZoomPanLayout logic.
Anyway, this is important that we have something functional to start with, as its easier to polish core functionalities and later add "optional" ones.
I'll do a code review shortly (probably tonight), and let you know.

**Comment by moagrius** [comment]
Thanks Peter!

**Comment by peterLaurence** [comment]
Just to let you know that i had some time tonight, but it wasn't enough. I hope to finish the job tomorrow.

**Comment by moagrius** [comment]
No rush at all!  Take your time

**Comment by peterLaurence** [comment]
So here are my remarks :

As per [awakenScrollBars](http://grepcode.com/file/repository.grepcode.com/java/ext/com.google.android/android/5.1.1_r1/android/view/View.java#View.awakenScrollBars%28int%2Cboolean%29) documentation, "this method calls invalidate() so the caller should not call invalidate()". This is done in `smoothScrollBy` though.
But, looking at android ScrollView, `awakenScrollBars` isn't called at this point. I believe that this isn't necessary because at this point we are just setting the starting and ending values of the scroll.

That leads to another point. It looks like they made an optimization in `computeScroll` : they don't call `scrollTo` as they have to deal with overscroll and other stuff that `scrollTo` doesn't. But most importantly, `scrollTo` calls `awakenScrollBars` at the end. This why (as they explain in the source of `computeScroll`) they replicate most of `scrollTo` code and then call `awakenScrollBars`.
I guess this is what we will end up doing for the overscroll mode.

I see that in `onMeasure` you didn't call the super method. While our implementation is very simple, doesn't this makes our ScrollView specific a TileView usage? On the other hand, this may have an impact on performance if we have a more complex implementation.

As for the rest of the code (which is herited from `ZoomPanLayout`), i don't have remarks for now :)
I'm still looking at android ScrollView (and meditating on it).

**Comment by moagrius** [comment]
Great feedback, Peter.

I copied most of the blocks with `awakenScrollBars` directly from the oooold version of android.widget.ScrollView (from Froyo!) - happy to make any updates there you'd like.

Same thoughts about computeScroll - if there are optimizations to be made, let's do it - normally that's the one entrypoint we get into Scroller/OverScroller updates, so this is where we call scrollTo...  If there's a better way, let me know.

Great point about onMeasure.  I have to admit, I kind of gave up a little here - I notice the android ScrollViews don't do anything special - they check if fillViewport is enabled, and if so, they measure the 1 child, and make it the same height of the container if it's not already at least that large.  That's it.  You can see from the demos that it "works", but that's being generous.  When I leave onMeasure alone, then the child is only ever as large as the container - which obviously does no good at all.  So I'm just letting it measure itself as if there is no constraint, but I don't feel great about it - for example, what would happen when the child is set to MATCH_PARENT for width, and WRAP_CONTENT for height (which should be the most common implementation), and the rows are wider than the container?  I suspect it'd scroll on both axes, which is wrong.  I need to figure out what went wrong here, and I am finding no clues in the original source.

LMK if any of that doesn't make sense, or isn't clear.

Generally speaking, I think the best way to move forward is a new project.  I'll do an empty project, commit that to master, then add these files as a new branch with PR, that way you can make your own commits, changes, and line comments.

**Comment by peterLaurence** [comment]
> I copied most of the blocks with awakenScrollBars directly from the oooold version of android.widget.ScrollView (from Froyo!) - happy to make any updates there you'd like.

We agree that we should take care of old versions of android compatibility. But as long as we are not adding code specific to newer api level, i prefer looking at the latest version of ScrollView.
The latest version don't call at all awakenScrollBars (nor invalidate), but calls `postInvalidateOnAnimation`, which doesn't exists in api level 2. So what about using `ViewCompat.postInvalidateOnAnimation`?


> for example, what would happen when the child is set to MATCH_PARENT for width, and WRAP_CONTENT for height (which should be the most common implementation), and the rows are wider than the container? I suspect it'd scroll on both axes, which is wrong. I need to figure out what went wrong here, and I am finding no clues in the original source.

Maybe this is reason why such both axis ScrollView doesn't exist in official api. But maybe we can find a work around. For example, we could consider those different cases :

- a MATCH_PARENT on width and WRAP_CONTENT on height would have the same behavior/implementation as android ScrollView.
- a MATCH_PARENT on height and WRAP_CONTENT on width would have the same behavior/implementation as android HorizontalScrollView.
- a WRAP_CONTENT on both width and height would unlock the scrolling on both axis (using maybe the current implementation, e.g let the child measure itself).

> Generally speaking, I think the best way to move forward is a new project. I'll do an empty project, commit that to master, then add these files as a new branch with PR, that way you can make your own commits, changes, and line comments.

Good idea, i admit this isn't very handy at the moment to make comments.

**Comment by moagrius** [comment]
> So what about using ViewCompat.postInvalidateOnAnimation?

+1.  We can implement once I get this in it's own project, then you can make line comments or commit directly.

> Maybe this is reason why such both axis ScrollView doesn't exist in official api. But maybe we can find a work around. For example, we could consider those different cases :

A user should be able to supply any dimension and it should work - imagine TileView where we provide a fixed size.  I'm sure I'm just missing something, I'll try to figure it out next time I sit down to this project.

I'll ping you when it's up.

**Comment by moagrius** [comment]
Hey Peter,

Good news - I found how the children are being measured in the native ScrollView: https://github.com/android/platform_frameworks_base/blob/master/core/java/android/widget/ScrollView.java#L1302.  Looks like they're doing basically the same thing I did, and that we're going to have to do a more complicated version to support both axes - still shouldn't be a huge deal though, I thikn I have my head around what's required.  Once I get that worked out and committed (hopefully in the next day or two), I'll post as a new project.

**Comment by peterLaurence** [comment]
Nice! I'm really interested to see what you will get, to learn from that. I'm not as comfortable as i should in this area (the onMeasure thing, and how to deal with children, etc).

**Comment by moagrius** [comment]
Updated in the new repo - LMK if you don't get invites / links /etc.

**Comment by moagrius** [comment]
Both `ScrollView` and `ScalingScrollView` are completely decomposed from `TileView` in V3, and I encourage anyone to use them independently of `TileView`.  Closing.
